### ISSUE TITLE
Two comets merge/sync error when sending file with large history footprint

### ISSUE THREAD

**Issue by mikolajpp** [urbit/arvo#594]
Scenario: 
2 comets, one merging desk from another. 
|merge (or |sync) results in periodic errors as shown below:

```
recover: dig: over
over
[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.636 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.637 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.638 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.639 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.640 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.641 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.642 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.643 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.643 14].[2.643 42]>
~tamnul_micwyn:dojo/plat|«ames»
recover: dig: over
over
[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.636 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.637 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.638 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.639 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.640 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.641 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.642 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.643 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.643 14].[2.643 42]>
~tamnul_micwyn:dojo/plat|«ames»
recover: dig: over
over
[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.636 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.637 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.638 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.639 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.640 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.641 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.642 9].[2.643 43]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.643 9].[2.643 43]>
```

**Comment by cgyarvin** [comment]
That's very strange, it looks to be crashing in ++jam -- as though we're
stack overflowing simply because we're sending an object that's too big.
Do you have a reliable reproduction case?

**Comment by mikolajpp** [comment]
Unfortunately, I garbage collected this comet. 

If this is due to a large object, I will try to reproduce: the large object could be history for 'app.js' which is packed plato app. It weighs ~1MB with debugging enabled, and I put it every time I rebuild it.

**Comment by mikolajpp** [comment]
Ok, here is what I did:

1. Have 2 comets
2. Pack one comet with history of a single large file ~1MB. (I saved it around 7 times)
3. Watch errors:
```

[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.675 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 12].[2.676 45]>
~mopber_malmus:dojo/plato> /«ames»
recover: dig: over
over
[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.675 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 12].[2.676 45]>
> 
> 
> 
> 
~mopber_malmus:dojo/plato> \«ames»
recover: dig: over
over
[%swim-call-vane %a %want ~]
/~zod/home/git-eaea2b528c/sys/zuse:<[2.675 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 7].[2.676 46]>
/~zod/home/git-eaea2b528c/sys/zuse:<[2.676 12].[2.676 45]>

```

Hopefully this is the same. If not, I can try to stack my comet even more to see.

**Comment by mikolajpp** [comment]
Now, even when I |cancel merge on receiving end, the errors on source continue. 
Not gonna garbage this comet and keep it for repro if you can't, but it seems it became unusable. It must be stuck in some "pack data" loop.

The errors can be stopped when I take the receiving comet offline. 
Once it comes up again, the source comet once again starts sending stuff and spits these errors.

**Comment by cgyarvin** [comment]
Problem is almost certainly the big history, which we're trying to sync.  I
bet there's more head-recursion traps like this in Urbit...

**Comment by mikolajpp** [comment]
Yes, I remember when I spawned fresh comet which had only one instance of said fail, 
the merge worked. No idea what is the threshold, but after 7 copies it starts failing.

**Comment by cgyarvin** [comment]
Can you work around it?  Not saying this isn't a low threshold, but your
data may still be excessive...

**Comment by mikolajpp** [comment]
Well, I can :-)
I think even if this was fixed, putting app into history every time it was recompiled is not a good idea. After all, I don't keep under git either. 

The reason I have done this, is because I don't want to mess with CORS while app makes requests to Urbit from different address/port. So I will keep stuffing my development comet to the brim, but will sync from another comet which does not have app.js in history. 

Another thing is, this will need a good solution wrt Urbit package distribution system. I can definitely imagine a package having 100 releases/bugfixes which will distributed through Urbit. 
I gather it is technically possible to request "latest" revision using clay: future package manager will need to take this into account. When I do 'pacman -S linux' it does not download all the historic releases from the last 20 years ;-)

**Comment by mikolajpp** [comment]
I have been thinking about this. Certain things should never be kept
under true revision control. Nobody(?) does that that for binary data, where diff operation is not what we usually mean by saying 'meaningful'. 

These are versioned in a human versioning system, usually categorized by time or features.

One thing that is evident, is that allowing pruning from the level of Urbit is a security issue: it should not be possible to lose data (unless it was originally marked as such) even if user executes malicious hoon. 

Another way of looking at this, is by trying to force things which do not belong in the revision control, we break our revision control. It should be possible to prune log/memory, but only from outside of Urbit, as a "housekeeping" tool.

I am sure this is already well understood by Tlonians ;-)

Possible solutions come to mind:

1. Allow eyre to serve from 'loom', i.e. not from clay. 
    Awkward and I am not sure how that would work. What about paths, permissions, etc.

2. A new vane similar to clay, except that much simpler and no 'diff' versioning. Cool, but what about 
    integration and operability? Maybe reuse stuff from clay, I would need to understand arvo more to know ;-)

3. A new kind of node in %clay, which would allow for pruning (and pruning only it)

I can think of many usecases where users will definitely want to opt-out from version control: confidential/sensitive data, pastebin like stuff which they would nonetheless want to be accessible from eyre, and so on. I haven't dig into clay, so no idea which of these are workable, but at some point we will need this.

**Comment by cgyarvin** [comment]
You have basically just described the correct division between state in a
%gall app and %clay data.

The correct answer is that %eyre needs to get better at serving data out of
%gall apps.  You can help us design this!

%gall is metadata -- only the current state should be kept.  %clay is data
-- a full history should be kept.

Obviously it's tempting to think of %clay as somehow more persistent than
%gall.  This is just a bad hangover from the 20th century.

**Comment by mikolajpp** [comment]
This is a good point about artificial feeling that clay = persistence. I catch myself thinking about stuffing stuff in clay several times a day ; )

Basically in this view application UI indeed should be made part of state? Which makes sense: no one has ever said "I want to use outdated GUI",  at least in a sane world.

This of course will need more thinking, but having Eyre to hook into Gall apps through some unified interface to retrieve "UI" is what I want and feels correct.

Interesting thing to solve will be "seeding an app": obviously ui blobs/HTML cannot be embedded inside source code (or we are back to square one), but must be injected after app starts. We could imagine various interesting scenarios, where we can inject data into app from eyre, some package management scheme could source UI according to app version from repository of Hoon apps etc.

Let's say I will do UP3 first, in the meantime hopefully ~palfun will finish clay work, which put together could be a start to designing some general solution to %gall - %eyre interface except the usual pokes subscriptions

**Comment by joemfb** [comment]
I believe this issue is the same as the "infinite loop" of stack traces described in #740, and fixed in #756. @mikolajpp, can you try to to reproduce this on the release-candidate branches of both arvo and vere?

**Comment by jtobin** [comment]
Closing as related issues seem to have been resolved, and this hasn't been followed up on since.
